CBinding/Gui/CTextEditorExtension.cs

```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
-				if (lineText.EndsWith(pair.Key)) {
-					lineText = lineText.Substring (0, lineText.Length - pair.Key.Length);
-					
-					int nameStart = lineText.LastIndexOfAny (allowedCharsMinusColon) + 1;
-					string itemName = lineText.Substring (nameStart).Trim ();
-					
-					if (string.IsNullOrEmpty (itemName))
-						return null;
-					
-					return Task.FromResult ((ICompletionDataList)pair.Value (this, pair.Key, itemName));
+			ICompletionDataList list = new CompletionDataList ();
+			if (shouldCompleteOn(completionChar)) {
+				CProject project = DocumentContext.Project as CProject;
+				unsavedFiles = new List<CXUnsavedFile> ();
+				foreach (Document openDocument in IdeApp.Workbench.Documents) {
+					if (openDocument.IsDirty) {
+						CXUnsavedFile unsavedFile = new CXUnsavedFile ();
+						unsavedFile.Filename = openDocument.FileName;
+						unsavedFile.Length = openDocument.Editor.Text.Length;
+						unsavedFile.Contents = openDocument.Editor.Text;
+						if (project.BOMPresentInFile [openDocument.FileName]) {
+							unsavedFile.Length += 3;
+							unsavedFile.Contents = "   " + unsavedFile.Contents;
+						}
+						unsavedFiles.Add (unsavedFile);
+					}
 				}
+				string operatorPattern = "operator\\s*(\\+|\\-|\\*|\\/|\\%|\\^|\\&|\\||\\~|\\!|\\=|\\<|\\>|\\(\\s*\\)|\\[\\s*\\]|new|delete)";
+				bool fieldOrMethodMode = completionChar == '.' || completionChar == '>' ? true : false;
+				IntPtr pResults = project.cLangManager.codeComplete (completionContext, unsavedFiles.ToArray (), DocumentContext.Name);
+				CXCodeCompleteResults results = Marshal.PtrToStructure<CXCodeCompleteResults> (pResults);
+				if (results.Results.ToInt64 () != 0) {
+					for(int i = 0; i < results.NumResults; i++) {
+						IntPtr iteratingPointer = results.Results + i * Marshal.SizeOf<CXCompletionResult>();
+						CXCompletionResult resultItem = Marshal.PtrToStructure<CXCompletionResult> (iteratingPointer);
+						CXCompletionString completionString = new CXCompletionString(resultItem.CompletionString);
+						uint completionchunknum = clang.getNumCompletionChunks (completionString.Pointer);
+						for (uint j = 0; j < completionchunknum; j++) {
+							if (clang.getCompletionChunkKind(completionString.Pointer, j) != CXCompletionChunkKind.CXCompletionChunk_TypedText)
+								continue;
+							uint priority = clang.getCompletionPriority (completionString.Pointer);
+							CXString cxstring = clang.getCompletionChunkText(completionString.Pointer, j);
+							string realstring = Marshal.PtrToStringAnsi (clang.getCString(cxstring));
+							clang.disposeString (cxstring);
+							if (resultItem.CursorKind == CXCursorKind.CXCursor_Destructor
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_UnaryOperator
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_BinaryOperator
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_CompoundAssignOperator
+								|| System.Text.RegularExpressions.Regex.IsMatch (realstring, operatorPattern)
```

Comment:
Compiling and retaining the regex instance is better than recreating it every time.